In [14]:
import os

working_dir = r"/content/fr-data"

In [21]:
import nltk
from nltk.tokenize import word_tokenize
nltk.download('punkt')
import pymorphy2
import string

morph = pymorphy2.MorphAnalyzer()

def preprocessing(working_dir):
  corpus = []

  for root, dirs, files in os.walk(working_dir):
    for name in files:
      fpath = os.path.join(root, name)
      with open(fpath, 'r') as f:  
        text = word_tokenize(f.read())
        text = [w.lower() for w in text]
        table = str.maketrans('', '', string.punctuation)
        text = [word for word in text if word.isalpha()]
        text = [morph.parse(word)[0].normal_form for word in text]
        corpus.append(' '.join(text))
        
  
  return corpus
        


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


In [22]:
pip install pymorphy2[fast]

In [24]:
corpus = preprocessing(working_dir)
#corpus[:10]

In [25]:
from sklearn.feature_extraction.text import CountVectorizer
import numpy as np

vectorizer = CountVectorizer()

In [26]:
X = vectorizer.fit_transform(corpus)

#vectorizer.get_feature_names()

In [27]:
matrix_freq = np.asarray(X.sum(axis=0)).ravel()
final_matrix = np.array([np.array(vectorizer.get_feature_names()), matrix_freq])


Берем список всех слов. Для каждого делаем в словаре список номеров документов, в кот. они встречаются. Дальше просто считаем количество элементов и

In [30]:
msw = ''
freq_s = 0
freq_l = 100
mlw = ''
words = vectorizer.get_feature_names()

for word in words:
  ind = vectorizer.vocabulary_.get(word)
  if ind > freq_s:
    msw = word
    freq_s = ind
  if ind < freq_l:
    mlw = word
    freq_l = ind

print("{}: {}".format(msw, freq_s))
print("{}: {}".format(mlw, freq_l))

ёще: 14554
after: 0


**Ответ на вопрос 1:** *ещё* (индекс 14554)
**Ответ на вопрос 2:** *after* (индекс 0, первое по списку, вероятно)

In [52]:
char_names = ["моника", "рэйчел", "чендлер", "росс", "фиби", "джоу"]

mpn = ''
freq = 0

for name in char_names:
  ind = vectorizer.vocabulary_.get(name)
  if ind > freq:
    mpn = name
    freq = ind
    print("{}: {}".format(mpn, freq))


моника: 5873
рэйчел: 10982
чендлер: 13933


In [57]:
import pandas as pd

original_df = pd.DataFrame(final_matrix)
original_df.to_pickle("./fr_data.pkl")

**Ответ на вопрос 4:** Чендлер (индекс 13933)